test-commons/src/main/java/org/sonarsource/analyzer/commons/checks/verifier/internal/QuickFixVerifier.java

```diff
@@ -0,0 +1,342 @@
+import java.util.stream.Collectors;
+import org.sonarsource.analyzer.commons.checks.verifier.quickfix.QuickFix;
+import org.sonarsource.analyzer.commons.checks.verifier.quickfix.TextEdit;
+import org.sonarsource.analyzer.commons.checks.verifier.quickfix.TextSpan;
+
+public class QuickFixVerifier implements Consumer<Set<InternalIssue>> {
+
+  private static final Pattern ISSUE_DEFINITION = Pattern.compile("\\s*quickfixes=(?<id>\\S+)\\]\\].*+");
+  private static final Pattern QUICK_FIX_MESSAGE = Pattern.compile("\\s*fix@(?<id>\\S++)\\s++\\{\\{(?<message>.*)\\}\\}");
+  private static final Pattern QUICK_FIX_EDIT = Pattern.compile("\\s*edit@(?<id>\\S+).++");
+  private static final Pattern ISSUE_LINE_MODIFIER = Pattern.compile("\\s*Noncompliant@(?<mod>\\S+)\\s*\\[\\[");
+
+  private static final UnaryOperator<String> VALUE_PATTERN = str -> str + "(?<value>[^;]+).+";
+  private static final Pattern START_LINE = Pattern.compile(VALUE_PATTERN.apply("\\s*sl="));
+  private static final Pattern END_LINE = Pattern.compile(VALUE_PATTERN.apply("\\s*el="));
+  private static final Pattern START_COLUMN = Pattern.compile(VALUE_PATTERN.apply("\\s*sc="));
+  private static final Pattern END_COLUMN = Pattern.compile(VALUE_PATTERN.apply("\\s*ec="));
+  private static final String PROPS_START_DELIMITER = "[[";
+  private static final String PROPS_END_DELIMITER = "]]";
+  public static final String VALUE_GROUP = "value";
+
+  private final Map<IssueIdentifier, List<QuickFix>> expectedQuickFixes = new HashMap<>();
+  private final Map<IssueIdentifier, List<String>> quickFixesForIssue = new HashMap<>();
+
+  //in these two maps we put the messages and edits, not caring if the qf definition was already found or not
+  private final Map<String, String> quickfixesMessages = new HashMap<>();
+  private final Map<String, List<RelativeTextEdit>> quickfixesEdits = new HashMap<>();
+  //this is used to calculate absolute line numbers for text edits
+  private final Map<String, Integer> quickfixesLineReference = new HashMap<>();
+
+  public QuickFixVerifier(List<Comment> expectedQuickFixesComments) {
+    buildExpectedQuickFixes(expectedQuickFixesComments);
+  }
+
+  @Override
+  public void accept(Set<InternalIssue> issues) {
+    for (InternalIssue issue : issues) {
+      IssueIdentifier issueIdentifier = getIssueIdentifier(issue);
+      List<QuickFix> actual = issue.quickFixes;
+      List<String> expectedQfId = quickFixesForIssue.getOrDefault(issueIdentifier, null);
+      if (expectedQfId == null) {
+        // We don't have to always test quick fixes, we do nothing if there is no expected quick fix.
+        continue;
+      }
+      List<QuickFix> expected = expectedQuickFixes.get(issueIdentifier);
+      validateQuickfixes(expected, actual, issue);
+      expectedQuickFixes.remove(issueIdentifier);
+    }
+    if (!expectedQuickFixes.isEmpty()) {
+      throw new AssertionError("[Quick Fix] Missing quick fixes for the following issues: " +
```

Comment:
What about something like `[Quick Fix] Expected but could not find quick fixes for the following issues: "`? To me it makes it more clear that the quick fix is missing from _actual_ and not from _expected_.